events.v2/events_test.go

```diff
@@ -0,0 +1,93 @@
+package events_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	events "github.com/tendermint/tmlibs/events.v2"
+)
+
+func TestSubscribe(t *testing.T) {
+	t.Parallel()
+
+	s := events.NewServer(1)
+	s.Start()
+	defer s.Stop()
+
+	ch := s.Subscribe("")
+	s.Publish("Goblin Queen")
+	assert.Equal(t, "Goblin Queen", <-ch)
+}
+
+func TestQueries(t *testing.T) {
+	t.Parallel()
+
+	s := events.NewServer(1)
+	s.Start()
+	defer s.Stop()
+
+	ch1 := s.Subscribe("tm.events.type=NewBlock")
+	s.PublishWithTags("Iceman", map[string]interface{}{"tm.events.type": "NewBlock"})
+	assert.Equal(t, "Iceman", <-ch1)
+
+	ch2 := s.Subscribe("tm.events.type=NewBlock AND abci.account.name=Igor")
+	s.PublishWithTags("Iceman", map[string]interface{}{"tm.events.type": "NewBlock", "abci.account.name": "Igor"})
+	assert.Equal(t, "Iceman", <-ch2)
+
+	ch3 := s.Subscribe("tm.events.type=NewBlock AND abci.account.name=Igor")
+	s.PublishWithTags("Iceman", map[string]interface{}{"tm.events.type": "NewBlock"})
+	<-ch1
+	assert.Empty(t, ch3)
+}
+
+func TestPublish(t *testing.T) {
+	t.Parallel()
+
+	s := events.NewServer(1)
+	s.Start()
+	defer s.Stop()
+
+	ch := s.Subscribe("")
+	s.Publish("Ka-Zar")
+	s.Publish("Ka-Zar")
+	assert.Equal(t, "Ka-Zar", <-ch)
+	assert.Equal(t, "Ka-Zar", <-ch)
+}
+
+func TestUnsubscribe(t *testing.T) {
+	t.Parallel()
+
+	s := events.NewServer(1)
+	s.Start()
+	defer s.Stop()
+
+	ch := s.Subscribe("")
+	s.Publish("Lester")
+	assert.Equal(t, "Lester", <-ch)
```

Comment:
write a wrapper using `select` with timeout = 10ms so we don't block if a value is not there